src/cr/cube/slices.py

```diff
@@ -0,0 +1,1140 @@
+# encoding: utf-8
+
+from __future__ import division
+
+import numpy as np
+from scipy.stats.contingency import expected_freq
+
+from cr.cube.util import lazyproperty
+from cr.cube.enum import DIMENSION_TYPE as DT
+
+
+class _CatXCatSlice(object):
+    def __init__(self, dimensions, counts, base_counts):
+        self._dimensions = dimensions
+        self._counts = counts
+        self._base_counts = base_counts
+
+    @lazyproperty
+    def _row_dimension(self):
+        return self._dimensions[0]
+
+    @lazyproperty
+    def _column_dimension(self):
+        return self._dimensions[1]
+
+    @lazyproperty
+    def _row_elements(self):
+        return self._row_dimension.valid_elements
+
+    @lazyproperty
+    def _column_elements(self):
+        return self._column_dimension.valid_elements
+
+    @lazyproperty
+    def _row_generator(self):
+        return zip(self._counts, self._base_counts, self._row_elements)
+
+    @lazyproperty
+    def _column_generator(self):
+        return zip(self._counts.T, self._base_counts.T, self._column_elements)
+
+    @lazyproperty
+    def rows(self):
+        return tuple(
+            _CategoricalVector(counts, base_counts, element.label, self.table_margin)
+            for counts, base_counts, element in self._row_generator
+        )
+
+    @lazyproperty
+    def columns(self):
+        return tuple(
+            _CategoricalVector(counts, base_counts, element.label, self.table_margin)
+            for counts, base_counts, element in self._column_generator
+        )
+
+    @lazyproperty
+    def table_margin(self):
+        return np.sum(self._counts)
+
+
+class _1DCatSlice(_CatXCatSlice):
+    @lazyproperty
+    def columns(self):
+        return tuple(
+            [
+                _CategoricalVector(
+                    self._counts, self._base_counts, "Summary", self.table_margin
+                )
+            ]
+        )
+
+
+class _MrXCatSlice(_CatXCatSlice):
+    @lazyproperty
+    def rows(self):
+        """Use only selected counts."""
+        return tuple(
+            _CategoricalVector(counts[0], base_counts[0], element.label, table_margin)
+            for counts, base_counts, element, table_margin in self._row_generator
+        )
+
+    @lazyproperty
+    def _row_generator(self):
+        return zip(
+            self._counts, self._base_counts, self._row_elements, self.table_margin
+        )
+
+    @lazyproperty
+    def columns(self):
+        """Use bother selected and not-selected counts."""
+        return tuple(
+            _MultipleResponseVector(
+                counts, base_counts, element.label, self.table_margin
+            )
+            for counts, base_counts, element in self._column_generator
+        )
+
+    @lazyproperty
+    def table_margin(self):
```

Comment:
We need this at slice level, to be able to assign it to vectors, so each vector (either row or col) can know _all_ that it needs to know in order for _any_ of its measures. I'm planning for this to also happen with zscores and pvals.

Reply:
If this works, let's run with it and sort it out in refactoring. The general idea I have here I mentioned in another comment, which is to produce it at the slice level when we need it, perhaps by calling a method on row-vectors that passes col-vectors to it. I don't have my head in the code enough yet to say specifically, but the general principle here is that a calculation responsibility should be on the object that has the data to perform it. That avoids passing it around to other objects, which has a strong tendency to grow without bound.

Anyway, another item we can address in refactoring.